pkg/core/combat/select.go

```diff
@@ -0,0 +1,279 @@
+package combat
+
+import (
+	"sort"
+
+	"github.com/genshinsim/gcsim/pkg/core/geometry"
+)
+
+// all targets
+
+func enemiesWithinAreaFiltered(a AttackPattern, filter func(t Enemy) bool, originalEnemies []Target) []Enemy {
+	var enemies []Enemy
+	hasFilter := filter != nil
+	for _, v := range originalEnemies {
+		e, ok := v.(Enemy)
+		if !ok {
+			panic("enemies should contain targets that implement the Enemy interface")
+		}
+		if hasFilter && !filter(e) {
+			continue
+		}
+		if !v.IsAlive() {
+			continue
+		}
+		if !e.IsWithinArea(a) {
+			continue
+		}
+		enemies = append(enemies, e)
+	}
+	return enemies
+}
+
+func gadgetsWithinAreaFiltered(a AttackPattern, filter func(t Gadget) bool, originalGadgets []Gadget) []Gadget {
+	var gadgets []Gadget
+	hasFilter := filter != nil
+	for _, v := range originalGadgets {
+		if v == nil {
+			continue
+		}
+		if hasFilter && !filter(v) {
+			continue
+		}
+		if !v.IsAlive() {
+			continue
+		}
+		if !v.IsWithinArea(a) {
+			continue
+		}
+		gadgets = append(gadgets, v)
+	}
+	return gadgets
+}
```

Comment:
nit: can simplify this down to a single helper function with generics (I think, not a go expert)
```go
func withinAreaFiltered[T Gadget | Enemy](a AttackPattern, filter func(t T) bool, original []Target) []T {
	var out []T
	hasFilter := filter != nil
	for _, v := range original {
		if v == nil {
			continue
		}
		if hasFilter && !filter(v) {
			continue
		}
		if !v.IsAlive() {
			continue
		}
		if !v.IsWithinArea(a) {
			continue
		}
		out= append(out, v)
	}
	return out
}
```

you could technically do the same for the handler methods, but you'd need to if/else based on `T` type since you need to call `c.gadgets` or `c.enemies`, thought it would be good if we can minimize the amount of copy/pasted logic in the other functions

Reply:
you are not allowed to use interfaces that contain methods in a union, so `Gadget | Enemy` is not legal